lualib/lua_scanners/avast_rest.lua

```diff
@@ -0,0 +1,354 @@
+    end
+
+    local upstream = rule.upstreams:get_upstream_round_robin()
+    local addr = upstream:get_addr()
+    local retransmits = rule.retransmits
+
+    local request_data = {
+      task = task,
+      timeout = rule.timeout,
+    }
+
+    local filename
+    if rule.use_files then
+      filename =  string.format('%s/%s.tmp',
+          rule.tmpdir, rspamd_util.random_hex(32))
+      local message_fd = rspamd_util.create_file(filename)
+
+      if not message_fd then
+        rspamd_logger.errx('cannot store file for avast_rest scan: %s', filename)
+        return
+      end
+
+      if type(content) == 'string' then
+        -- Create rspamd_text
+        local rspamd_text = require "rspamd_text"
+        content = rspamd_text.fromstring(content)
+      end
+      content:save_in_file(message_fd)
+
+      -- Ensure cleanup
+      task:get_mempool():add_destructor(function()
+        os.remove(filename)
+        rspamd_util.close_file(message_fd)
+      end)
+
+      request_data.headers = {
+        ['X-Correlation-Id'] = string.sub(task:get_uid(), 1,6)
+      }
+
+      request_data.url = make_url(addr, filename)
+    else
+
+      if rule.scan_mime_parts then
+
+        local _, _, mime_attr = maybe_part:get_type_full()
+        local det_ext = maybe_part:get_detected_ext()
+
+        if mime_attr.name then
+          local ext,_ = common.gen_extension(mime_attr.name)
+          if ext then
+            filename = string.format('file.%s', ext)
+          end
+        elseif det_ext then
+          filename = string.format('file.%s', det_ext)
+        else
+          filename = "file.txt"
+        end
+      else
+        filename = "file.eml"
+      end
+      request_data.url = make_url(addr, filename)
+      request_data.headers = {
+        ['Content-Type'] = 'application/octet-stream',
+        ['X-Correlation-Id'] = string.sub(task:get_uid(), 1,6)
+      }
+      request_data.body = content
+    end
+
+    lua_util.debugm(rule.name, task, '%s: request: url: %s, headers: %s, size: %s',
+      rule.log_prefix, request_data.url, request_data.headers, #request_data.body)
+
+    local function avast_rest_callback(http_err, code, body, headers)
+
+      local function requery()
+        -- set current upstream to fail because an error occurred
+        upstream:fail()
+
+        -- retry with another upstream until retransmits exceeds
+        if retransmits > 0 then
+
+          retransmits = retransmits - 1
+
+          lua_util.debugm(rule.name, task,
+              '%s: Request Error: %s - retries left: %s',
+              rule.log_prefix, http_err, retransmits)
+
+          -- Select a different upstream!
+          upstream = rule.upstreams:get_upstream_round_robin()
+          addr = upstream:get_addr()
+          request_data.url = make_url(addr, filename)
+
+          lua_util.debugm(rule.name, task, '%s: retry IP: %s:%s',
+              rule.log_prefix, addr, addr:get_port())
+
+          http.request(request_data)
```

Comment:
I'd also suggest to add `upstream` to `request_data` as it will simplify handling. You might also need to remove all calls to `upstream:ok` and `upstream:fail` as that will be handled internally by `lua_http` module.